amplipi/streams.py

```diff
@@ -1321,6 +1323,224 @@ def info(self) -> models.SourceInfo:
+
+    src_config_folder = f"{utils.get_folder('config')}/srcs/v{self.src}"
+    loc = f'{src_config_folder}/currentSong'
+    try:
+      with open(loc, 'r', encoding='utf-8') as file:
+        data = json.loads(file.read())
+        self.ended = data['state'] == 'ENDED'
+    except Exception:
+      pass
+
+    if self.state == 'playing' and self.playing in self.song_list and self.ended and self.song_index < len(self.song_list) - 1:
+      self.next_song()
+    elif self.ended and (self.song_index >= len(self.song_list) - 1 or self.playing not in self.song_list):
+      self.state = 'paused'
+
+  def next_song(self):
+    self.change_song(self.song_index + 1)
+
+  def previous_song(self):
+    self.change_song(self.song_index - 1)
+
+  def change_song(self, new_song_id):
+    self.song_index = new_song_id
+    self.playing = self.song_list[self.song_index]
+    self.restart()
+
+    f = open(self.command_file_path, 'w')
+    f.write('play')
+    f.close()
+    self._prev_timeout = datetime.datetime.now() + datetime.timedelta(seconds=3)
+    self.ended = False
+
+  def send_cmd(self, cmd):
+    if cmd in self.supported_cmds:
+      if cmd == 'stop':
+        self._deactivate()
+      if self.command_file_path is not None:
+        if cmd == 'pause':
+          f = open(self.command_file_path, 'w')
+          f.write('pause')
+          f.close()
+          self.state = 'paused'
+
+        if cmd == 'play':
+          if not self._is_running():
+            logger.info(f'running: {self.vlc_args}')
+            self.proc = subprocess.Popen(args=self.vlc_args, preexec_fn=os.setpgrp)
+          f = open(self.command_file_path, 'w')
+          f.write('play')
+          f.close()
+          self.state = 'playing'
+
+        if cmd == 'next':
+          self.next_song()
+
+        if cmd == 'prev':
+          # Restart current song if we are past the first 3 seconds, otherwise go back
+
+          if self._prev_timeout > datetime.datetime.now() and self.song_index > 0:
+            self.previous_song()
+          else:
+            self.change_song(self.song_index)
+
+  def info(self) -> models.SourceInfo:
+    self.supported_cmds = ['play', 'pause', 'prev']
+    if self.song_index != len(self.song_list) - 1 and self.playing in self.song_list:
+      self.supported_cmds.append('next')
+
+    img = 'static/imgs/no_note.png'
+    if self.playing != '':
+      img = 'static/imgs/note.png'
+    source = models.SourceInfo(name=self.full_name(),
+                               state=self.state,
+                               img_url=img,
+                               supported_cmds=self.supported_cmds,
+                               type=self.stream_type)
+    src_config_folder = f"{utils.get_folder('config')}/srcs/v{self.src}"
+    loc = f'{src_config_folder}/currentSong'
+    try:
+      with open(loc, 'r', encoding='utf-8'):
+        source.track = self.playing.split('/')[-1]
+    except Exception:
+      pass
+    return source
```

Comment:
As we discussed IRL the next track playback depends on the system being polled. Shouldn't this happen in a separate thread or process?

Reply:
Disregard this for now.